torch/_inductor/fx_passes/post_grad.py

```diff
@@ -650,7 +650,6 @@ def reinplace_scatters(graph):
             ):
                 src = src.args[0]
             copy_args_to_copy_nodes[(dst, src)] = node
-            assert node.args[0].op == "placeholder"
```

Comment:
hi @Chillee, I encountered a case where the first arg op was not a `placeholder` for `aten.copy_.default`,  which caused the assertion error.
```
graph():
   ...
   %full_default_2 : [num_users=2] = call_function[target=torch.ops.aten.full.default](args = ([4, 16, 2, 2], 0), kwargs = {dtype: torch.float32, layout: torch.strided, device: cpu, pin_memory: False})
   %where_self : [num_users=1] = call_function[target=torch.ops.aten.where.self](args = (%logical_not_default, %scalar_tensor_default, %full_default_2), kwargs = {})
   %copy__default : [num_users=1] = call_function[target=torch.ops.aten.copy_.default](args = (%full_default_2, %where_self), kwargs = {})
   ...
```

 Is it good to modify the code like this?

Reply:
@Valentine233 This... shouldn't be possible, I think. Is this a graph you're getting from `torch.compile`? cc: @bdhirsh

Reply:
Yes, it is from `torch.compile`.

For more details, I paste the python code and the whole graph here:
```
def dot_prod_attention(
    query: torch.Tensor, key: torch.Tensor, value: torch.Tensor
) -> torch.Tensor:
    """Input tensors assumed to have shape (batch_size, seq_len, n_head, embed_dim)"""
    q = query.transpose(1, 2)
    k = key.transpose(1, 2)
    v = value.transpose(1, 2)
    bs = q.size(0)
    k_len = k.size(-2)
    attn_mask = torch.ones(
        bs, k_len, dtype=torch.bool, device=query.device
    ).tril(diagonal=0)
    q = q / 3.0
    scores = torch.matmul(q, k.transpose(-2, -1))
    attn_mask = (attn_mask == 0).view((bs, 1, 1, k_len)).expand_as(scores)
    scores = scores.masked_fill(attn_mask, -float("inf"))
    weights = torch.nn.functional.softmax(scores, dim=-1)
    return torch.matmul(weights, v)
```
```
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %arg2_1 : [num_users=1] = placeholder[target=arg2_1]
    %full_default_1 : [num_users=0] = call_function[target=torch.ops.aten.full.default](args = ([], -inf), kwargs = {dtype: torch.float32, layout: torch.strided, device: cpu, pin_memory: False})
    %permute_default : [num_users=1] = call_function[target=torch.ops.aten.permute.default](args = (%arg0_1, [0, 2, 1, 3]), kwargs = {})
    %permute_default_1 : [num_users=1] = call_function[target=torch.ops.aten.permute.default](args = (%arg1_1, [0, 2, 1, 3]), kwargs = {})
    %permute_default_2 : [num_users=1] = call_function[target=torch.ops.aten.permute.default](args = (%arg2_1, [0, 2, 1, 3]), kwargs = {})
    %iota : [num_users=1] = call_function[target=torch.ops.prims.iota.default](args = (2,), kwargs = {start: 0, step: 1, dtype: torch.int64, device: cpu, requires_grad: False})
    %unsqueeze : [num_users=1] = call_function[target=torch.ops.aten.unsqueeze.default](args = (%iota, -2), kwargs = {})
    %iota_1 : [num_users=1] = call_function[target=torch.ops.prims.iota.default](args = (4,), kwargs = {start: 0, step: 1, dtype: torch.int64, device: cpu, requires_grad: False})
    %unsqueeze_1 : [num_users=1] = call_function[target=torch.ops.aten.unsqueeze.default](args = (%iota_1, -1), kwargs = {})
    %sub : [num_users=1] = call_function[target=torch.ops.aten.sub.Tensor](args = (%unsqueeze, %unsqueeze_1), kwargs = {})
    %le : [num_users=1] = call_function[target=torch.ops.aten.le.Scalar](args = (%sub, 0), kwargs = {})
    %full_default : [num_users=1] = call_function[target=torch.ops.aten.full.default](args = ([4, 2], True), kwargs = {dtype: torch.bool, layout: torch.strided, device: cpu, pin_memory: False})
    %logical_and : [num_users=1] = call_function[target=torch.ops.aten.logical_and.default](args = (%le, %full_default), kwargs = {})
    %eq_scalar : [num_users=1] = call_function[target=torch.ops.aten.eq.Scalar](args = (%logical_and, 1), kwargs = {})
    %view_default : [num_users=1] = call_function[target=torch.ops.aten.reshape.default](args = (%eq_scalar, [4, 1, 1, 2]), kwargs = {})
    %expand_default : [num_users=1] = call_function[target=torch.ops.aten.expand.default](args = (%view_default, [4, 16, 2, 2]), kwargs = {})
    %logical_not_default : [num_users=1] = call_function[target=torch.ops.aten.logical_not.default](args = (%expand_default,), kwargs = {})
    %scalar_tensor_default : [num_users=1] = call_function[target=torch.ops.aten.scalar_tensor.default](args = (-inf,), kwargs = {dtype: torch.float32, layout: torch.strided, device: cpu})
    %full_default_2 : [num_users=2] = call_function[target=torch.ops.aten.full.default](args = ([4, 16, 2, 2], 0), kwargs = {dtype: torch.float32, layout: torch.strided, device: cpu, pin_memory: False})
    %where_self : [num_users=1] = call_function[target=torch.ops.aten.where.self](args = (%logical_not_default, %scalar_tensor_default, %full_default_2), kwargs = {})
    %copy__default : [num_users=1] = call_function[target=torch.ops.aten.copy_.default](args = (%full_default_2, %where_self), kwargs = {})
    %mul_scalar : [num_users=1] = call_function[target=torch.ops.aten.mul.Scalar](args = (%permute_default, 0.5773502691896257), kwargs = {})
    %permute_default_3 : [num_users=1] = call_function[target=torch.ops.aten.permute.default](args = (%permute_default_1, [0, 1, 3, 2]), kwargs = {})
    %mul_scalar_1 : [num_users=1] = call_function[target=torch.ops.aten.mul.Scalar](args = (%permute_default_3, 0.5773502691896257), kwargs = {})
    %expand_default_1 : [num_users=1] = call_function[target=torch.ops.aten.expand.default](args = (%mul_scalar, [4, 16, 2, 32]), kwargs = {})
    %clone_default : [num_users=1] = call_function[target=torch.ops.aten.clone.default](args = (%expand_default_1,), kwargs = {memory_format: torch.contiguous_format})
    %view_default_1 : [num_users=1] = call_function[target=torch.ops.aten.reshape.default](args = (%clone_default, [64, 2, 32]), kwargs = {})
    %expand_default_2 : [num_users=1] = call_function[target=torch.ops.aten.expand.default](args = (%mul_scalar_1, [4, 16, 32, 2]), kwargs = {})
    %clone_default_1 : [num_users=1] = call_function[target=torch.ops.aten.clone.default](args = (%expand_default_2,), kwargs = {memory_format: torch.contiguous_format})
    %view_default_2 : [num_users=1] = call_function[target=torch.ops.aten.reshape.default](args = (%clone_default_1, [64, 32, 2]), kwargs = {})
    %bmm_default : [num_users=1] = call_function[target=torch.ops.aten.bmm.default](args = (%view_default_1, %view_default_2), kwargs = {})
    %view_default_3 : [num_users=1] = call_function[target=torch.ops.aten.reshape.default](args = (%bmm_default, [4, 16, 2, 2]), kwargs = {})
    %add_tensor : [num_users=2] = call_function[target=torch.ops.aten.add.Tensor](args = (%view_default_3, %copy__default), kwargs = {})
    %amax_default : [num_users=1] = call_function[target=torch.ops.aten.amax.default](args = (%add_tensor, [-1], True), kwargs = {})
    %sub_tensor : [num_users=1] = call_function[target=torch.ops.aten.sub.Tensor](args = (%add_tensor, %amax_default), kwargs = {})
    %exp_default : [num_users=2] = call_function[target=torch.ops.aten.exp.default](args = (%sub_tensor,), kwargs = {})
    %sum_dim_int_list : [num_users=1] = call_function[target=torch.ops.aten.sum.dim_IntList](args = (%exp_default, [-1], True), kwargs = {})
    %div_tensor : [num_users=1] = call_function[target=torch.ops.aten.div.Tensor](args = (%exp_default, %sum_dim_int_list), kwargs = {})
    %expand_default_3 : [num_users=1] = call_function[target=torch.ops.aten.expand.default](args = (%div_tensor, [4, 16, 2, 2]), kwargs = {})
    %view_default_4 : [num_users=1] = call_function[target=torch.ops.aten.reshape.default](args = (%expand_default_3, [64, 2, 2]), kwargs = {})
    %expand_default_4 : [num_users=1] = call_function[target=torch.ops.aten.expand.default](args = (%permute_default_2, [4, 16, 2, 32]), kwargs = {})
    %clone_default_2 : [num_users=1] = call_function[target=torch.ops.aten.clone.default](args = (%expand_default_4,), kwargs = {memory_format: torch.contiguous_format})
    %view_default_5 : [num_users=1] = call_function[target=torch.ops.aten.reshape.default](args = (%clone_default_2, [64, 2, 32]), kwargs = {})
    %bmm_default_1 : [num_users=1] = call_function[target=torch.ops.aten.bmm.default](args = (%view_default_4, %view_default_5), kwargs = {})
    %view_default_6 : [num_users=1] = call_function[target=torch.ops.aten.reshape.default](args = (%bmm_default_1, [4, 16, 2, 32]), kwargs = {})
    return (view_default_6,)
```

Reply:
@Valentine233 can you provide full repro code? Seems very weird

Reply:
Sure. You could pull this branch, restore `post_grad.py` and run `python test/inductor/test_fused_attention.py -k test_sdpa_rewriter_15_cpu`.

The error is: 
```
File "/workspace/pytorch/torch/_inductor/compile_fx.py", line 390, in compile_fx_inner
    compiled_graph = fx_codegen_and_compile(
  File "/workspace/pytorch/torch/_inductor/compile_fx.py", line 583, in fx_codegen_and_compile
    post_grad_passes(gm, is_inference=is_inference)
  File "/workspace/pytorch/torch/_inductor/fx_passes/post_grad.py", line 95, in post_grad_passes
    reinplace_scatters(gm.graph)
  File "/workspace/pytorch/torch/_inductor/fx_passes/post_grad.py", line 653, in reinplace_scatters
    assert node.args[0].op == "placeholder"
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
AssertionError:

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
```

Reply:
@Chillee Hi, do you have any findings about this?

Reply:
This was fixed by https://github.com/pytorch/pytorch/pull/115808.